test/other/testMemPoolTypes.cpp

```diff
@@ -0,0 +1,129 @@
+#include "System/MemPoolTypes.h"
+#include "System/Log/ILog.h"
+
+#include <vector>
+#include <cstddef>
+
+#define CATCH_CONFIG_MAIN
+#include "lib/catch.hpp"
+
+namespace {
+struct TestData
+{
+    TestData(uint64_t& counter) :
+        counter(++counter) { }
+    ~TestData() {
+        --counter;
+    }
+    uint64_t& counter;
+};
+
+template <typename T>
+struct AllocFixture {
+    AllocFixture() : mempool(std::make_unique<T>()) {}
+
+    TestData* alloc() {
+        return mempool->template alloc<TestData>(instanceCounter);
+    }
+
+    void free(TestData* obj) {
+        return mempool->free(obj);
+    }
+
+    std::unique_ptr<T> mempool;
+    uint64_t instanceCounter=0;
+};
+
+TEMPLATE_TEST_CASE_METHOD(AllocFixture, "test allocator when full", "[class][template]",
+        (StaticMemPool<1,sizeof(TestData)>),
+        (FixedDynMemPool<sizeof(TestData), 1, 1>))
+{
+    AllocFixture<TestType> inst;
+    auto& mempool = *inst.mempool;
+    auto* obj = inst.alloc();
+    REQUIRE(obj);
+    REQUIRE(!mempool.can_alloc());
+
+    // StaticMemPool silently fails
+    //auto* obj2 = mempool.allocMem(1);
+    //REQUIRE(!obj2);
+}
+
+TEMPLATE_TEST_CASE_METHOD(AllocFixture, "test allocator base functionality", "[class][template]",
+        (StaticMemPool<1024,sizeof(TestData)>),
+        (FixedDynMemPool<sizeof(TestData), 1, 1024>))
+{
+    AllocFixture<TestType> inst;
+    auto& mempool = *inst.mempool;
+
+    REQUIRE(mempool.can_alloc());
+
+    std::vector<TestData*> allocated;
+
+    for (size_t i =0; i < mempool.NUM_PAGES(); ++i) {
+        REQUIRE(mempool.can_alloc());
+        auto obj = inst.alloc();
+        REQUIRE(mempool.alloced(obj));
+        allocated.push_back(obj);
+    }
+
+    REQUIRE(inst.instanceCounter == mempool.NUM_PAGES());
+
+    REQUIRE(!mempool.can_alloc());
+
+    for (auto* obj : allocated) {
+        REQUIRE(mempool.can_free());
+        mempool.free(obj);
+    }
+    REQUIRE(inst.instanceCounter == 0);
+    REQUIRE(!mempool.can_free());
+}
+
+TEMPLATE_TEST_CASE_METHOD(AllocFixture, "test reuse allocator's memory", "[class][template]",
+        (StaticMemPool<1024,sizeof(TestData)>),
+        (FixedDynMemPool<sizeof(TestData), 1, 1024>))
+{
+    AllocFixture<TestType> inst;
+    auto& mempool = *inst.mempool;
+
+    std::vector<TestData*> allocated;
+    for (size_t i =0; i < mempool.NUM_PAGES(); ++i) {
+        auto obj = inst.alloc();
+        allocated.push_back(obj);
+    }
+
+    std::for_each(allocated.rbegin(), allocated.rend(), [&](auto* obj) {
+        mempool.free(obj);
+    });
+
+    // verify that new elements reuse previous memory
+    for (size_t i = 0; i < mempool.NUM_PAGES(); ++i) {
+        auto obj = inst.alloc();
+        REQUIRE(obj == allocated[i]);
```

Comment:
This not only checks that the returned memory is the same, but also in the same order. Is that actually a constraint? For example maybe somebody could optimize the mem pool but it would involve making it work like a stack so pages are returned like {1, 2, 3} the first time but {3, 2, 1} when reused. Perhaps turn `std::vector<TestData*> allocated` into `std::set`?

Reply:
I agree that this is not a functional requirement - I added a relevant comment to make it clear. I still find this test very helpful as this helps testing the implementation detail of the allocator. Without it we could not prove that the intention of reusing last entry (e.g. for cache reasons) is fullfilled.
Other allocators that do not obey the order do not need to execute this test. 
I will take a look at the other allocators as I have not covered every class in this PR.

Reply:
added separate test with std::map with counter